### PR TITLE
K8s 1.30 provider slim s390x

### PR DESCRIPTION
**Pull request by chandramerla**
**What this PR does / why we need it**:
Enables s390x support in kubernetes provider 1.30.
This is required for running s390x based tests spinning a k8s cluster using k8s provider.

**Special notes for your reviewer**:
Earlier PR (created for 1.28 provider) whose comments are addressed in this PR: #1201 
Changes required in respective prow jobs to publish provider are in PR: [#3566](https://github.com/kubevirt/project-infra/pull/3566) 

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by chandramerla** [comment]
I've run publish.sh inside the containers used in kubevirtci prow jobs, which build and publish gocli, centos9, k8s provider images and alpine container image. With the changes in publish.sh script in this PR and prow job related changes (In PR https://github.com/kubevirt/project-infra/pull/3566), I've successfully tested publishing of multiarch manifest-list for all the above images to icr.io (as I don't have write access in quay.io). In published multiarch manifest-lists s390x arch based images are built/present for gocli, centos9 and k8s-1.30 (slim).

**Comment by chandramerla** [comment]
@brianmcarey : Could you rerun tests and review this?

File: cluster-provision/centos9/scripts/vm.sh
Comment (brianmcarey) [inline cluster-provision/centos9/scripts/vm.sh:255]: This feels strange - we first create the `qemu_system_cmd` and then we remove the secondary network devices? Could we just not include them when creating `qemu_system_cmd` ?

File: cluster-provision/gocli/cmd/utils/images.go
Comment (brianmcarey) [inline cluster-provision/gocli/cmd/utils/images.go:7]: Thanks for this but why is this needed for s390x ? We should try to minimize our changes here as the PR is already quite big.

File: cluster-provision/gocli/cmd/utils/utils.go
Comment (brianmcarey) [inline cluster-provision/gocli/cmd/utils/utils.go:38]: I don't think this comment is needed here.

File: cluster-provision/k8s/1.30/extra-pre-pull-images
Comment (brianmcarey) [inline cluster-provision/k8s/1.30/extra-pre-pull-images:7]: Could the bump of calico be split out to a different PR? is this needed for the s390x provider?

File: cluster-provision/k8s/1.30/k8s_provision.sh
Comment (brianmcarey) [inline cluster-provision/k8s/1.30/k8s_provision.sh:69]: Is this still true? I saw there were recent crio releases?

File: cluster-provision/k8s/1.30/provision.sh
Comment (brianmcarey) [inline cluster-provision/k8s/1.30/provision.sh:65]: Would there be much needed to get this in the default repos?

File: cluster-up/check.sh
Comment (brianmcarey) [inline cluster-up/check.sh:59]:
```suggestion
	echo "[ERR ] $KVM_HPAGE KVM hugepage enabled. It needs to be disabled while nested virtualization is enabled for s390x"
```

File: publish.sh
Comment (brianmcarey) [inline publish.sh:7]: This is probably not needed.

**Review by brianmcarey — COMMENTED** [review]
thanks for this @chandramerla  - I have a couple of comments and questions.

File: cluster-provision/k8s/1.30/k8s_provision.sh
Comment (chandramerla) [reply cluster-provision/k8s/1.30/k8s_provision.sh:69]: Not sure if still true, I've tried a month back. So, let me once try with latest available crio [v1.30.5](https://github.com/cri-o/cri-o/releases/tag/v1.30.5) (or probably even I'll with v1.31.0 if 1.30 doesn't work)

File: cluster-provision/k8s/1.30/provision.sh
Comment (chandramerla) [reply cluster-provision/k8s/1.30/provision.sh:65]: I'm not sure. But, I will try finding how to do this and whom to contact, etc. If you know whom to reach, please let me know.

File: publish.sh
Comment (chandramerla) [reply publish.sh:7]: Removed

File: cluster-up/check.sh
Comment (chandramerla) [reply cluster-up/check.sh:59]: Thanks, corrected!

**Review by chandramerla — COMMENTED** [review]
@brianmcarey I've replied to your comments/questions.

File: cluster-provision/k8s/1.30/k8s_provision.sh
Comment (chandramerla) [reply cluster-provision/k8s/1.30/k8s_provision.sh]: reverted

File: cluster-provision/k8s/1.30/provision.sh
Comment (chandramerla) [reply cluster-provision/k8s/1.30/provision.sh:65]: I've asked the question in fedora-devel group https://matrix.to/#/!DMptnKvPTdgoULzUTV:fedoraproject.org/$rbqZqn1e4e5pIJ_UAnjr5xnJJor4c5M3RDFF-fMU5sk?via=fedoraproject.org&via=fedora.im&via=matrix.org

Comment (chandramerla) [reply cluster-provision/k8s/1.30/provision.sh:65]: @brianmcarey : I checked with community, basically there are no builds created by [CentOS nfv SIG](https://sigs.centos.org/nfv/) community for s390x. So, better I build from [sources](https://git.centos.org/rpms/openvswitch/branches?branchname=c8s-sig-nfv-openvswitch-2.16) (or better from [srpm](https://mirror.stream.centos.org/SIGs/9-stream/nfv/source/openvswitch-2/Packages/o/)), which has spec file to build an rpm.

Comment (chandramerla) [reply cluster-provision/k8s/1.30/provision.sh:65]: The current place from where I'm downloading is also from authentic user releng (it is fedora's release engineering), so I'm thinking may be as a follow-up PR I can fix building it from sources, if you are ok. Otherwise also let me know I will try building rpm from sources. I still need to figure out what exact commands/flags need to pass to rpm-build/moc, to be consistent with that of x86.

File: cluster-provision/k8s/1.30/extra-pre-pull-images
Comment (brianmcarey) [reply cluster-provision/k8s/1.30/extra-pre-pull-images:7]: The calico bump should be handled in a separate PR - as we will need to do this for all providers to keep it consistent.

File: cluster-provision/k8s/1.30/provision.sh
Comment (brianmcarey) [reply cluster-provision/k8s/1.30/provision.sh:65]: Yes I followed the conversation over on the fedora matrix - this should be ok for now.

File: cluster-provision/k8s/1.30/extra-pre-pull-images
Comment (chandramerla) [reply cluster-provision/k8s/1.30/extra-pre-pull-images:7]: ok. I'll start a separate PR for calico bump. 
Just to confirm my understanding - we keep (or try to keep)  same version of calico across all provider versions and update it across provider versions when necessary. Isn't it?

Comment (brianmcarey) [reply cluster-provision/k8s/1.30/extra-pre-pull-images:7]: Yes unless there is a specific reason to have an older version of calico on a certain provider but I don't think there is any reason in this case

File: cluster-provision/k8s/1.30/manifests/local-volume.yaml
Comment (brianmcarey) [inline cluster-provision/k8s/1.30/manifests/local-volume.yaml:116]: Can we apply this bump to all three providers in a separate PR?

**Comment by brianmcarey** [comment]
@chandramerla can you create an optional presubmit `
check-provision-k8s-1.30-s390x` so that we can test this?

File: cluster-provision/k8s/1.30/extra-pre-pull-images
Comment (chandramerla) [reply cluster-provision/k8s/1.30/extra-pre-pull-images]: Done. create separate PR https://github.com/kubevirt/kubevirtci/pull/1281/files

File: cluster-provision/k8s/1.30/manifests/local-volume.yaml
Comment (chandramerla) [reply cluster-provision/k8s/1.30/manifests/local-volume.yaml:116]: @brianmcarey a new PR created #1285

**Comment by chandramerla** [comment]
> @chandramerla can you create an optional presubmit ` check-provision-k8s-1.30-s390x` so that we can test this?

@brianmcarey Created the PR https://github.com/kubevirt/project-infra/pull/3681/files

**Comment by chandramerla** [comment]
> @chandramerla I've added the following base image with your changes: `quay.io/kubevirtci/centos9:2410230922-f2ef2f72`

Thanks, @brianmcarey . I've Updated the image in the cluster-provision/k8s/base-image. 
Could you please run all jobs again incl. check-provision-k8s-1.30-s390x ?

**Review by brianmcarey — COMMENTED** [review]
Looks good @chandramerla  - just a  couple of questions on the publish.

File: publish.sh
Comment (chandramerla) [reply publish.sh:114]: Thanks for pointing this.

I think this can be fixed as below - When this script run on x86 first, we are re-tagging only multi-arch manifest. Next, when run on s390x, s390x-arch image will be retagged.
```
  for i in ${IMAGES_TO_RETAG[@]}; do
    echo "INFO: retagging $i (previous tag $PREV_KUBEVIRTCI_TAG)"
    if [ $ARCH == "amd64" ]; then 
      skopeo copy "docker://${TARGET_REPO}/k8s-$i:${PREV_KUBEVIRTCI_TAG}" "docker://${TARGET_REPO}/k8s-$i:${KUBEVIRTCI_TAG}"
      skopeo copy "docker://${TARGET_REPO}/k8s-$i:${PREV_KUBEVIRTCI_TAG}-slim" "docker://${TARGET_REPO}/k8s-$i:${KUBEVIRTCI_TAG}-slim"
    elif [ $ARCH == "s390x" ] && [ $i == "1.30" ]; then
      skopeo copy "docker://${TARGET_REPO}/k8s-$i:${PREV_KUBEVIRTCI_TAG}-${ARCH}" "docker://${TARGET_REPO}/k8s-$i:${KUBEVIRTCI_TAG}-${ARCH}"
      skopeo copy "docker://${TARGET_REPO}/k8s-$i:${PREV_KUBEVIRTCI_TAG}-slim-${ARCH}" "docker://${TARGET_REPO}/k8s-$i:${KUBEVIRTCI_TAG}-slim-${ARCH}"
    fi
  done
```

**Review by chandramerla — COMMENTED** [review]
Thanks Brian for the review. I've replied to your comments.

**Review by brianmcarey — COMMENTED** [review]
/approve

thanks @chandramerla  - can we get a followup PR as soon as possible to apply these changes to the k8s-1.31 provider too so that they can be automatically included when we create the k8s-1.32 provider shortly?

File: cluster-provision/gocli/cmd/run.go
Comment (dhiller) [inline cluster-provision/gocli/cmd/run.go:583]: Could we introduce constants that tell the reader what this is about?

i.e.

```suggestion
const (
    QEMU_DEVICE_S390X  = "virtio-net-ccw"
    QEMU_DEVICE_X86_64 = "virtio-net-pci"
)
```

and use these? Or create a func similar to `GetSSHUserForArchitecture`?

File: cluster-provision/gocli/cmd/run.go
Comment (dhiller) [inline cluster-provision/gocli/cmd/run.go:604]: see above, then this would be more obvious.

File: cluster-up/check.sh
Comment (dhiller) [inline cluster-up/check.sh:37]: please use correct indentation (tabs)

File: cluster-up/cluster/ephemeral-provider-common.sh
Comment (dhiller) [inline cluster-up/cluster/ephemeral-provider-common.sh:203]: This shouldn't be necessary - iiuc the default for cdi deployment is flipped inside kubevirt/kubevirt: https://github.com/kubevirt/kubevirt/blob/0e54154a8a95745aa7cd54ccd03597c37b1269ed/hack/cluster-up.sh#L20

File: publish.sh
Comment (dhiller) [inline publish.sh:61]: When will this be removed? After this PR has landed?

**Review by dhiller — COMMENTED** [review]
@chandramerla I have added a couple of questions inline.

FYI @brianmcarey

File: cluster-provision/gocli/cmd/utils/utils.go
Comment (chandramerla) [reply cluster-provision/gocli/cmd/utils/utils.go:38]: Actually when i've written this function, pkg/libssh/ssh.go didn't exist. But later I didn't realise that we can move it here. Thanks for pointing this, now I've moved it accordingly.

File: cluster-provision/gocli/cmd/provision.go
Comment (chandramerla) [reply cluster-provision/gocli/cmd/provision.go:236]: Yes, removed the log

File: cluster-provision/gocli/cmd/run.go
Comment (chandramerla) [reply cluster-provision/gocli/cmd/run.go:583]: Done

File: cluster-provision/gocli/cmd/run.go
Comment (chandramerla) [reply cluster-provision/gocli/cmd/run.go:604]: Yes, updated.

File: cluster-up/check.sh
Comment (chandramerla) [reply cluster-up/check.sh:37]: Done

File: cluster-up/cluster/ephemeral-provider-common.sh
Comment (chandramerla) [reply cluster-up/cluster/ephemeral-provider-common.sh:203]: Yes, with changes from this [commit](https://github.com/kubevirt/kubevirtci/commit/22e4513dd963abd45332f2972e3021423b699d6d#diff-12e7b6e3d4ce92e7d50b6cd0e3aebac95e9480cb4dbbe4a220e1c86c3b12f6ebL124), I don't need to handle this here anymore. 
P.S: Earlier when discussing with Ammar, he suggested to change it here than changing the default in the run.go as that was the default in the shell scripts before he migrated.

Thanks for pointing this.

File: publish.sh
Comment (chandramerla) [reply publish.sh:61]: I've removed this in [PR](https://github.com/kubevirt/kubevirtci/pull/1316) for 1.31 slim provider supporting s390x.

**Comment by chandramerla** [comment]
> @chandramerla I have added a couple of questions inline.
> 
> FYI @brianmcarey

Thanks @dhiller for your comments/questions. Addressed all of them.

File: cluster-provision/gocli/cmd/run.go
Comment (dhiller) [inline cluster-provision/gocli/cmd/run.go:977]: please use tabs for indentation

**Review by dhiller — CHANGES_REQUESTED** [review]
@chandramerla nearly there - once the indentation has slipped, everything else looks fine!

File: cluster-provision/gocli/cmd/run.go
Comment (chandramerla) [reply cluster-provision/gocli/cmd/run.go]: Thanks. Actually I've couple of more places, where I fixed this indentation issue after you pointed this. I will take of such issues in future.

Comment (dhiller) [reply cluster-provision/gocli/cmd/run.go]: Not sure why spaces slippped in here - I remember that gofmt emits tabs as default

**Review by dhiller — APPROVED** [review]

**Comment by dhiller** [comment]
@chandramerla would you mind taking a look at the failing check-provision lanes please?

File: cluster-provision/gocli/cmd/run.go
Comment (chandramerla) [reply cluster-provision/gocli/cmd/run.go]: right. Not sure, but today I've (re-)installed all gotools and now I see it is working.

**Comment by chandramerla** [comment]
> @chandramerla would you mind taking a look at the failing check-provision lanes please?

@dhiller Thanks for the labels.
I'm seeing all the check-provision jobs (even [outside](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/1317/check-provision-k8s-1.29/1853650013293383680) this PR) are failing with same kubevirt conformance tests related to sig-network.

The common error across all those tests is

> /usr/bin/virt-chroot create-tap --tap-name tap0 --uid 107 --gid 107 --queue-number 0 --mtu 1480, err: exit status 1, output: Error: failed to create tap device named tap0. Reason: [0]: operation not permitted

This is happening since last evening/night, before that I've not seen those sig-network failures.
I don't have much clue here. So I was trying to reach @brianmcarey  on the same.

**Comment by dhiller** [comment]
> > @chandramerla would you mind taking a look at the failing check-provision lanes please?
> 
> @dhiller Thanks for the labels. I'm seeing all the check-provision jobs (even [outside](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirtci/1317/check-provision-k8s-1.29/1853650013293383680) this PR) are failing with same kubevirt conformance tests related to sig-network.
> 
> The common error across all those tests is
> 
> > /usr/bin/virt-chroot create-tap --tap-name tap0 --uid 107 --gid 107 --queue-number 0 --mtu 1480, err: exit status 1, output: Error: failed to create tap device named tap0. Reason: [0]: operation not permitted
> 
> This is happening since last evening/night, before that I've not seen those sig-network failures. I don't have much clue here. So I was trying to reach @brianmcarey on the same.

@EdDev @ormergi do you have an idea on why this is happening - why is the tap getting a not permitted error?

**Comment by dhiller** [comment]
What we noticed is that the last change to kubevirtci happened on 29th Oct, where the failure started happening yesterday.

**Comment by dhiller** [comment]
So it might seem some change on kubevirt itself is responsible here.

**Comment by EdDev** [comment]
> > The common error across all those tests is
> > > /usr/bin/virt-chroot create-tap --tap-name tap0 --uid 107 --gid 107 --queue-number 0 --mtu 1480, err: exit status 1, output: Error: failed to create tap device named tap0. Reason: [0]: operation not permitted
> > 
> > 
> > This is happening since last evening/night, before that I've not seen those sig-network failures. I don't have much clue here. So I was trying to reach @brianmcarey on the same.
> 
> @EdDev @ormergi do you have an idea on why this is happening - why is the tap getting a not permitted error?

I do not know what has changed to trigger this, but I do not think it is related to the k/k project.

The permission problem can be caused by SELinux or chroups. These are the reasons I saw in the past.
I suggest you run this one on the kubevirt/kubevirt project.

**Comment by ormergi** [comment]
I noticed the new provider use generic cloud image instead of vagrant box for the cluster node OS
Maybe its worth trying with vagrant box to eliminate this is the issue?

Also, I saw the kernel args are slight diffrent, for example the s390x provider kernel args doesnt specify `nopti`, maybe it case some slowness when tap device is created when selinux is involved.
https://github.com/kubevirt/kubevirtci/blob/5481bd92fa3804fad0f14821cda8aff5ae209f24/cluster-provision/centos9/scripts/kernel.s390x.args
https://github.com/kubevirt/kubevirtci/blob/5481bd92fa3804fad0f14821cda8aff5ae209f24/cluster-provision/centos9/scripts/kernel.args

**Comment by brianmcarey** [comment]
> I noticed the new provider use generic cloud image instead of vagrant box for the cluster node OS Maybe its worth trying with vagrant box to eliminate this is the issue?
> 
> Also, I saw the kernel args are slight diffrent, for example the s390x provider kernel args doesnt specify `nopti`, maybe it case some slowness when tap device is created when selinux is involved. https://github.com/kubevirt/kubevirtci/blob/5481bd92fa3804fad0f14821cda8aff5ae209f24/cluster-provision/centos9/scripts/kernel.s390x.args https://github.com/kubevirt/kubevirtci/blob/5481bd92fa3804fad0f14821cda8aff5ae209f24/cluster-provision/centos9/scripts/kernel.args

Hi @ormergi  - thanks for taking a look at this - we are seeing this on all kubevirtci PRs since the 4th. For example a simple PR - https://github.com/kubevirt/kubevirtci/pull/1317

**Comment by ormergi** [comment]
> Hi @ormergi - thanks for taking a look at this - we are seeing this on all kubevirtci PRs since the 4th. For example a simple PR - #1317

Thanks for the heads up
I see last change was centoOS bump https://github.com/kubevirt/kubevirtci/pull/1314, its not the first time we see issues with such changes, lets try to revert this change and see if things stabilize, WDTY?

**Comment by brianmcarey** [comment]
/hold 

To avoid needless retesting.

**Comment by chandramerla** [comment]
> Hi, since then the changes were populated to k8s versions > 1.30 ? This PR added changes to 1.30 only Thanks

Hi @oshoval 

Yes, apart from these changes for 1.30, some changes were done in 1.31 https://github.com/kubevirt/kubevirtci/pull/1316/files.
With that, all following versions like 1.32 (which made a copy from 1.31) have the changes.
Hope this answers your question.

**Comment by oshoval** [comment]
Hi @chandramerla 
Yep thank you

**Pull request merged**
